### PR TITLE
Midi 65/results

### PR DESCRIPTION
**Pull request by WojciechMat**
I have implemented a prototype streamlit app to compare predictions for whole pieces. 
I think I will later incorporate it into dashboard.py
It is very messy, just wanted to get it to work.

I will analyze it more tommorow.
image

### CODE REVIEW

**Comment by WojciechMat** [comment]
To train the model and see the results you can run
```
python train.py device=cuda:0

streamlit run --server.port 4466 dashboard.py
```
You will be able to choose which model you want to use and which index you want to start plotting samples from.
If you change display mode to Predict piece you will be able to choose which piece with model predicted velocity to render.

**Comment by WojciechMat** [comment]
I have uploaded some checkpoints to [huggingface](https://huggingface.co/wmatejuk/midi-transformer/tree/main).
You can download them to models/ directory and run dashboard to look through the results.

**Comment by WojciechMat** [comment]
### Dashboard updates
I have added table with model parameters to the dashboard:
image


You can also run dashboard with giant-midi-sustain dataset by running (I suggest using 10% of the dataset as it is really gigantic):
```
streamlit run --server.port 4466 dataset.dataset_name="roszcz/giant-midi-sustain" 'dataset_split="train[:10%]"'
```
Notice you have to put additional apostrophe quotation marks if you are using brackets in the parameter.
You have to set dataset_split to train as it is the only split in giant-midi-sustain dataset.

File: dashboard.py [hunk 1]
```diff
@@ -28,14 +35,18 @@ def get_sample_info(dataset: BinsToVelocityDataset, midi_filename: str):
...
 
-def model_predictions_review():
-    # options
-    path = st.selectbox(label="model", options=glob.glob("models/*.pt"))
-
-    start_index = eval(st.text_input(label="start index", value="0"))
+def model_predictions_review(cfg: DictConfig):
+    with st.sidebar:
+        # options
+        path = st.selectbox(label="model", options=glob.glob("models/*.pt"))
```
Comment (roszcz) [inline dashboard.py:42]: I think it would be better to operate on run names rather than model files - that way we could reference the results in wandb easily. 

Also, looks like right now models are saved only for runs that actually finished, but I think it would be better to save every time there is an improvement in validation loss:

```python
best_test_loss = float("inf")  # or float(cfg.min_save_loss)
for epoch in epochs:
    train_epoch()
    v_loss = val_epoch():
    if v_loss <= best_test_loss:
        save_checkpoint(model, optimizer, cfg)
        best_test_loss = v_loss
```

If you then use those checkpoints here, you'll be able to show the full training config (maybe at the bottom).

File: dashboard.py [hunk 2]
```diff
@@ -108,11 +123,12 @@ def model_predictions_review():
...
-        name = filename.split("/")[0] + "-" + str(idx + start_index) + "-"
-        pred_piece.source["midi_filename"] = train_cfg.run_name + "/" + name + os.path.basename(filename)
+        name = f"{filename.split('.')[0].replace('/', '-')}-{idx + start_index}-"
+        directory = "tmp/dashboard/"
+        pred_piece.source["midi_filename"] = model_dir + "/" + name + ".mid"
 
-        name = filename.split("/")[0] + "-" + str(idx) + "-qv-" + bins + "-"
-        quantized_vel_piece.source["midi_filename"] = "common/" + name + os.path.basename(filename)
+        name = f"{filename.split('.')[0].replace('/', '-')}-{idx + start_index}-qv-{bins}-{dataset.sequence_len}-"
+        quantized_vel_piece.source["midi_filename"] = directory + "common/" + name + ".mid"
```
Comment (roszcz) [inline dashboard.py]: In the maestro datasets we have the information about composer name and the piece title - we should make effort to display it here.

The approach I think should work for this, is to track additional field for every record with a dumped source information:

```python
record = {
    ...
    'source': json.dumps(piece.source)
}
```

This should be enough to deal with the fact, that this `source` is not unified between datasets. 

You should also be able to track the actual start and finish note indexes and display those on the dashboard as well.

File: dashboard.py [hunk 3]
```diff
@@ -1,41 +1,183 @@
...
+    print("Making predictions ...")
+    results = make_examples(dataset=dataset, model=model, start_index=start_index, n_examples=n_samples)
+
+    bins = train_cfg.dataset.bins.replace(" ", "-")
+    for it in range(n_samples):
+        # use every second record, so as not to create overlapped examples - it works together with make_examples()
+        idx = it * 2
+
+        src = results[it]["src"]
+        out = results[it]["out"]
```
Comment (roszcz) [inline dashboard.py]: I think this `make_examples` usage is a very awkward pattern, that you're forced into because of the `Batch` class existence :)

The way I think this could work:

```python
n_samples = 10
idxs = np.random.randint(len(dataset), size=n_samples)
records = dataset[idxs]

for record in records:
    model_output = process_record(record)
    piece_info = json.loads(record["source"])
    
    piece_gt = MidiPiece(record)

    piece_pred = MidiPiece(record)
    piece_pred.df.velocity = model_output.velocity
    ...
```

File: evals.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    idx = -1
+    for b in dataloader:
+        batch = Batch(b[0], b[1], pad_idx)
+        for it in range(len(batch)):
+            idx += 1
+            # I want to be able to get samples from any index from the database,
+            # I also want to use examples without overlap
+            if idx < start_index or idx % 2 == 1:
+                continue
+
```
Comment (roszcz) [inline evals.py]: Like I hinted in other comments, I think sample selection should be separated from model output generation

**Comment by WojciechMat** [comment]
I have implemented your suggestions to the dashboard. 
Samples for Model predictions now are chosen at random. I use a process_record function to make model output, for each record individually, in the form of tokenized sample.
Tables on dashboard now include composer and title information from piece source.
image

**Comment by roszcz** [comment]
@WojciechMat what's the status of this PR? I'm not sure if I should refactor this branch or master :)

If this is the most up-to-date with the velocity work, let's solve conflicts and merge this
